viewing_party/party.py

```diff
@@ -1,23 +1,270 @@
+    users_watched_movies = []
+
+    for movie in user_watched_list:
+        users_watched_movies.append(movie['title'])
+
+    return users_watched_movies
+
+
+def get_friends_unique_watched(user_data):
+    '''
+    collects unique watched movies from friends, stores them in a list,
+    and returns the list
+    '''
+    unique_watched_list = []
+
+    # User Data Watched Sets
+    friend_watched_set = set(get_friends_watched(user_data))
+    user_watched_set = set(get_users_watched(user_data))
+
+    unique_set = friend_watched_set - user_watched_set
+
+    # User Data Variables
+    friends_list_dict = user_data["friends"]
+
+
+    for i in range(len(friends_list_dict)):
+        for friend_movie in friends_list_dict[i]["watched"]:
+            if friend_movie["title"] in unique_set and friend_movie not in unique_watched_list:
+                unique_watched_list.append(friend_movie) 
+
+    return unique_watched_list
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    '''
+    Returns a list of recommended movies based on what
+    the user has a subscription to using user_data as input
+    '''
+    recommended_movies_list = []
+
+    # User Data Watched Sets
+    friend_watched_set = set(get_friends_watched(user_data))
+    user_watched_set = set(get_users_watched(user_data))
+    
+    unique_set = friend_watched_set - user_watched_set
+
+    # User Data Variables
+    subscriptions_list = user_data['subscriptions']
+    friends_list_dict = user_data["friends"]
+
+
+    # Pull data for recommended movies from user data
+    for i in range(len(friends_list_dict)):
+        for friend_movie in friends_list_dict[i]["watched"]:
+            if friend_movie["title"] in unique_set and friend_movie['host'] in subscriptions_list and friend_movie not in recommended_movies_list:
+                recommended_movies_list.append(friend_movie) 
+
+    return recommended_movies_list
+
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+# made it to wave 5!! *party emoji*
+
+def get_new_rec_by_genre(user_data):
+    '''
+    Returns a list of recommended movies that the user has not seen
+    based on the user's most watched genres
+    '''
+    recommended_movies_list = []
+
+    # User Data Watched Sets
+    friend_watched_set = set(get_friends_watched(user_data))
+    user_watched_set = set(get_users_watched(user_data))
+    
+    unique_set = friend_watched_set - user_watched_set
+
+    # User data Variables
+    friends_list_dict = user_data["friends"]
+    most_watched_genre = get_most_watched_genre(user_data)
+
+
+    # Pull data for recommended movies from user data
+    for i in range(len(friends_list_dict)):
+        if len(user_data['watched']) == 0 or len(friends_list_dict[i]['watched']) == 0:
+            break

```

Comment:
This isn't the best place for this check and there's a potential bug.  I recommend checking if `user_data["watched"]` is empty at the beginning of the function to short circuit the algorithm entirely and skip any unnecessary steps.  If friends_list_dict[i]['watched'] is empty, the loop that starts on line 227 will not run, so this friend dictionary will be skipped.  The potential bug would come in if a user has some friends with empty `watched` lists, and then this check would end this loop early